tc/autotuner/autotuner.h

```diff
@@ -0,0 +1,214 @@
+  std::mutex bestTimeMtx_;
+  std::mutex executorsMutex_;
+  std::atomic_bool stopRequested_;
+  std::atomic_size_t currentCompilationJob_;
+  std::atomic_size_t numEvaluations_;
+  std::queue<std::unique_ptr<typename Backend::ExecutorType>> executors_;
+  std::queue<CandidateConfiguration*> configurations_;
+
+  /// inputs
+  lang::TreeRef kTcTree_;
+  const MappingOptionsType kBaseMapping_;
+  const std::unordered_map<size_t, std::vector<const DLConstTensor*>> kInputs_;
+  std::unordered_map<size_t, std::vector<const DLTensor*>> outputs_;
+
+  // results
+  size_t bestTime_;
+  MappingOptionsType bestMappingOptions_;
+
+  // backing options cache
+  std::shared_ptr<OptionsCache<Backend>>& optionsCache_;
+};
+} // namespace detail
+
+/**
+ * An Autotuner provides the basic interface to run a SearchStrategy over a
+ * particular Backend.
+ * Under the hood it instantiates a TuningHarness<Backend> and a
+ * SearchStrategy. The Autotuner then calls TuningHarness<Backend>::run()
+ * on the SearchStrategy.
+ */
+template <typename Backend, typename SearchStrategy>
+class Autotuner {
+ public:
+  using BackendType = Backend;
+  using MappingOptionsType = typename BackendType::MappingOptionsType;
+  using OptionsCacheType = tc::autotune::OptionsCache<Backend>;
+
+  /// Parse and store the underlying tree representation for all the TC
+  /// functions provided.
+  explicit Autotuner(const std::string& tc);
+
+  /// Runs autotuning on the TC function tcEntryPoint.
+  /// This assumes input and output tensors of proper sizes have been
+  /// initialized on each GPU. For now this initialization is left to the
+  /// client of the Autotuner.
+  /// Tuning performs a search of template type SearchStrategy starting from
+  /// baseMapping.
+  ///
+  /// Alternatively an OptionsCache cacheFileName serialized path
+  /// can be specified from which the tuner recover multiple starting points
+  /// and preempt the baseMapping. This can be useful in a reinforcement
+  /// situation where short tunings are ran and their result cached
+  /// iteratively.
+  ///
+  /// Lastly a TuningParameterFixer function can be specified to limit the
+  /// search space (i.e. when certain parameters are known to be good/bad
+  /// independently on a particular TC).
+  ///
+  /// \return an optional MappingOptions
+  llvm::Optional<MappingOptionsType> tune(
+      const std::string& tcEntryPoint,
+      const std::unordered_map<size_t, std::vector<const DLConstTensor*>>&
+          inputs,
+      std::unordered_map<size_t, std::vector<const DLTensor*>>& outputs,
+      const typename Backend::MappingOptionsType& baseMapping,
+      const std::string& cacheFileName = "",
+      const TuningParameterFixer& fixedParams = TuningParameterFixer());
+
+ private:
+  std::map<std::string, lang::TreeRef> tcEntryPointMap_;
+  std::shared_ptr<OptionsCache<Backend>> optionsCache_;
+};
+
+/// Helper functions that need specializing for various backends.
```

Comment:
I'm not quite sure what these functions are supposed to do.  They return the value of the same type as one of their arguments, so looks like some sort of copying but why do we need the second argument then?

Reply:
This is a bottom-up refactoring, this legacy behavior is related to TuningConfigurations and will need to be cleaned up but I am punting atm.